### PR TITLE
Add in autoadvance event for nextUp

### PR DESCRIPTION
**Pull request by jnatalzia**
### This PR will...
Add a new event to the player api which denotes when a player auto-advances to the next video after the `next up` ui is displayed.

### Why is this Pull Request needed?
We would like to add an API event when the player advances to the next video on its own in the event that the related plugin is being used to set up playlists/recs.

### Are there any points in the code the reviewer needs to double check?
It would be good to ensure that the consolidation of the event firing makes sense. Instead of simple making them a single API event (`nextAdvance` or something similar) this approach makes the two different types of advancement individually trackable.

### Are there any Pull Requests open in other repos which need to be merged with this?
This can be merged with associated functionality in other PRs

#### Addresses Issue(s):
DATA948

### CODE REVIEW

File: src/js/controller/controller.js [hunk 1]
```diff
@@ -539,7 +539,9 @@ Object.assign(Controller.prototype, {
             }
 
             // It wasn't the last item in the playlist,
-            //  so go to the next one
+            //  so go to the next one and trigger an autoadvance event
+            const related = _api.getPlugin('related');
+            triggerAdvanceEvent(related, 'nextAutoAdvance');
             _next({ reason: 'playlist' });
```
Comment (johnBartos) [inline src/js/controller/controller.js]: Doesn't `reason: 'playlist'` imply an auto advance?

File: src/js/controller/controller.js [hunk 2]
```diff
@@ -662,18 +664,25 @@ Object.assign(Controller.prototype, {
...
-                        itemsShown: [ nextUp ],
-                        feedData: nextUp.feedData,
-                    });
-                }
-                related.next();
+            triggerAdvanceEvent(related, 'nextClick', () => related.next());
+        }
+
+        function triggerAdvanceEvent(related, evt, cb) {
+            if (!related) {
```
Comment (johnBartos) [inline src/js/controller/controller.js]: I think it makes more sense to not have `related` as an argument because you're never using `related` at your callsites, outside of passing it to this function.

File: src/js/controller/controller.js (hunk 1 above)
Comment (jnatalzia) [reply src/js/controller/controller.js]: Yeah but if I understand the setup correctly, the next event that would fire would be a `beforePlay` when the playlist item has already changed. This autoadvance event needs to track before the playlist switches, so triggering a separate event ensures we tracking the autoadvance at the correct time. See `feedClick` vs `relatedClick`

File: src/js/controller/controller.js (hunk 2 above)
Comment (jnatalzia) [reply src/js/controller/controller.js]: I agree. I passed it in the avoid doing the plugin lookup twice (the callback for `nextClick` requires a reference to the related plugin). Another option is passing the related plugin into the callback and that way we keep `related` specific to this function but that's a little hard to follow when reading through the code imo. How do you feel about just doing the lookup twice in the case of the `nextclick` handler?

File: src/js/controller/controller.js (hunk 1 above)
Comment (johnBartos) [reply src/js/controller/controller.js]: Gotcha

File: src/js/controller/controller.js (hunk 2 above)
Comment (johnBartos) [reply src/js/controller/controller.js]: Oh I see what you mean actually. Fine the way it is.

Comment (robwalch) [reply src/js/controller/controller.js]: You're not calling it twice, but it's repeated twice, instead of only being written and called once in `triggerAdvanceEvent`.

Comment (robwalch) [reply src/js/controller/controller.js]: oh i see. passing in `() => related.next()`

File: src/js/controller/controller.js (hunk 1 above)
Comment (robwalch) [reply src/js/controller/controller.js]: Is it critical that we track normal playlist progression? What about "beforePlaylistItem" for any playlist item change to track drop off? I'm OK with this but am very concerned that we're overloading the player with events for something that we want 'related' to track.
